tools-local/Microsoft.ML.CodeAnalyzer/ContractsCheckAnalyzer.cs

```diff
@@ -0,0 +1,248 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.ML.CodeAnalyzer
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class ContractsCheckAnalyzer : DiagnosticAnalyzer
+    {
+        // Detecting that a syntax call is actually on a particular method is computationally
+        // intensive, so once we detect that we're on Contracts methods, we put all the methods
+        // here.
+
+        private const string Category = "Contracts";
+
+        public static class NameofDiagnostic
```

Comment:
Do these nested classes need to be public?

Reply:
I wasn't quite sure about this one. I guess there's no real point in making the outside names accessible in this way, since you could just call get supported diagnostics anyway.